### PR TITLE
Add a First View introduction overlay to the Stats section

### PR DESCRIPTION
**Pull request by mjuhasz**
This PR adds a First View introduction overlay to the Stats section.

The First View introduction overlay is shown to users when they first visit the Stats section. 

screencapture at tue jul 12 15 24 30 edt 2016

Clicking the "Got it!" button will hide the First View and show the Stats section.

If the "Don't show this again" checkbox is checked, the First View will never been shown to the user again. If it is not checked, the First View will be shown whenever the user visits the Stats section again.

If the "Don't show this again" checkbox was checked, you can delete the `calypso_store` IndexedDB table in order to reset the Redux state and allow the First View to be shown again.

### Testing notes:

- All browsers (desktop and mobile) should work
- Either navigating to the stats section via the sidebar or directly accessing a /stats URL will show the first view
- Clicking "Got it!" should hide the first view
    - If the "Don't show this again" checkbox is checked, the first view will never reappear (even after a reload of Calypso)
    - If the "Don't show this again" checkbox is not checked, the first view will reappear the next time the stats section is viewed
- Navigating away to a different section and then back to stats without clicking "Got it!" should show the first view again

Support for other sections will be possible once #6651 is completed (Reduxifying `<Main>`).

Test live: https://calypso.live/?branch=add/stats-first-view

### CODE REVIEW

File: client/components/first-view/index.jsx [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+			<RootChild className={ classes }>
+				<ReactCSSTransitionGroup transitionName="first-view-transition"
+						component={ TransitionGroupComponent }
+						transitionEnter={ false } transitionEnterTimeout={ 0 }
+						transitionLeaveTimeout={ 250 }>
+					{ this.props.isVisible && (
+						<div key="content" className={ firstViewContentClasses }>
+							{ this.props.children }
+
+							<Button onClick={ this.hide }>{ this.translate( 'Got it!' ) }</Button>
```
Comment (ehg) [inline client/components/first-view/index.jsx]: Maybe add some `context` for the the translators?

File: client/components/first-view/index.jsx [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+							<div className={ firstViewHidePreferenceClasses }>
+								<label>
+									<input type="checkbox"
+											checked={ ! this.state.isEnabled }
+											onChange={ this.enableOrDisableNextTime } />
+									{ this.translate( 'Don\'t show this again' ) }
```
Comment (ehg) [inline client/components/first-view/index.jsx]: Minor, but I prefer double quotes over escaping characters.

File: client/components/first-view/index.jsx [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+export default connect(
+	( state ) => {
+		const sectionName = getSectionName( state );
+
+		return {
+			sectionName: sectionName,
+			isVisible: shouldViewBeVisible( state ),
+		};
+	},
+	dispatch => bindActionCreators( {
```
Comment (ehg) [inline client/components/first-view/index.jsx]: I don't think you need `bindActionCreators` here. `connect( …, { hideView } )` should suffice.

File: client/components/first-view/index.jsx [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+	},
+
+	allowPageScrolling() {
+		document.documentElement.classList.remove( 'no-scroll' );
+	}
+} );
+
+export default connect(
+	( state ) => {
+		const sectionName = getSectionName( state );
```
Comment (ehg) [inline client/components/first-view/index.jsx]: Can this be inlined in the returned object?

File: client/components/main/index.jsx [hunk 5]
```diff
@@ -3,10 +3,10 @@
 /**
  * External dependencies
  */
-var React = require( 'react' ),
```
Comment (ehg) [inline client/components/main/index.jsx]: Since you're not making any functional changes to this file yet, I wouldn't touch it.

File: client/my-sites/stats/stats-main/index.jsx [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+	);
+};
+
+export default connect(
+	( state ) => {
+		return {
+			isFirstViewVisible: shouldViewBeVisible( state ),
+		};
+	},
+	null
```
Comment (ehg) [inline client/my-sites/stats/stats-main/index.jsx]: A `null` argument shouldn't be needed here.

File: client/components/main/index.jsx (hunk 5 above)
Comment (mjuhasz) [inline client/components/main/index.jsx]: I agree, fixed.

We can do cleanup in a dedicated PR in order to keep this one smaller, and easier to review.

File: client/components/first-view/index.jsx (hunk 2 above)
Comment (mjuhasz) [inline client/components/first-view/index.jsx]: Agreed. It's easier to read that way. Fixed.

File: client/components/first-view/index.jsx (hunk 1 above)
Comment (mjuhasz) [inline client/components/first-view/index.jsx]: Good point, fixed!

File: client/components/first-view/index.jsx (hunk 3 above)
Comment (mjuhasz) [inline client/components/first-view/index.jsx]: Indeed! Fixed!

File: client/my-sites/stats/stats-main/index.jsx (hunk 6 above)
Comment (mjuhasz) [inline client/my-sites/stats/stats-main/index.jsx]: True. Fixed!

File: client/components/first-view/index.jsx (hunk 4 above)
Comment (mjuhasz) [inline client/components/first-view/index.jsx]: Absolutely! Fixed!

**Comment by mjuhasz** [comment]
@ehg Thank you for your comments, all addressed! Good suggestions!

**Comment by ehg** [comment]
> Navigating away to a different section and then back to stats without clicking "Got it!" should show the first view again

I had some problems with this. I got a FirstView on stats, then appended the Guided Tour `?tour=main` arg, to check for interactions. I went through the tour. I then visited all sites `/stats`, where there was no FirstView, but the content div was blurred. I then tried going to other stats sections, and found no first view.

Is there any easy way to reset my first view history, so I can try to reproduce and describe in a less confusing manner?

**Comment by mattsherman** [comment]
@ehg Thanks for the review and suggestions!

To reset the first view history, you can delete the `calypso_store` IndexedDB table to allow the First View to be shown again.

I'll look into trying to reproduce the scenario you mentioned.

**Comment by mattsherman** [comment]
> I had some problems with this. I got a FirstView on stats, then appended the Guided Tour ?tour=main arg, to check for interactions. I went through the tour. I then visited all sites /stats, where there was no FirstView, but the content div was blurred. I then tried going to other stats sections, and found no first view.

@ehg Good catch! I was able to reproduce this -- it had nothing to do with the Guided Tour. The first view simply wasn't being included in the stats overview render. Fixed!

**Comment by shaunandrews** [comment]
In Safari, I can uncheck the the checkbox — but then I can never click it again.

**Comment by shaunandrews** [comment]
I signed in as a brand new user, and visited my Stats page. Everything looked fine. I clicked the "Got it!" button to dismiss the First View. Then I refreshed my browser. Something like this loaded:

screen shot 2016-07-13 at 2 05 08 pm

Its a strange "half-loaded" state — notice the Site Card in the sidebar. During this state, the First View is shown, however, once Calypso is "fully" loaded, First View transitions out.

**Comment by jonathansadowski** [comment]
Hmm, good catch @shaunandrews ...  it feels like we just need to add `isSectionLoading( state )` to our `shouldViewBeVisible` selector.

**Comment by shaunandrews** [comment]
I pushed some design changes that look like this:

screen shot 2016-07-13 at 4 58 19 pm

I'm still not done with it (I gotta figure out what to do with the illustration), but I think using a card really improves the contrast and legibility of the text.

**Comment by mattsherman** [comment]
Okay, small update... the "half-loaded" state issue that @shaunandrews found only happens in Safari Private Mode. I believe this is because IndexedDB and localStorage are not available when in private mode.

**Comment by shaunandrews** [comment]
Here's where I ended up with the design changes:

screen shot 2016-07-14 at 11 21 40 am

**Comment by ehg** [comment]
:+1: from me, @mcsf could you have a quick look at the actionLog bits, to make sure we're not going to break anything with https://github.com/Automattic/wp-calypso/pull/6417 :)

File: client/state/ui/first-view/selectors.js [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+	return doesViewHaveFirstView( view ) && ! isViewDisabled;
+}
+
+export function wasViewHidden( state, view ) {
+	return -1 !== state.ui.firstView.hidden.indexOf( view );
+}
+
+export function switchedFromDifferentSection( state ) {
+	const section = state.ui.section;
+	const routeSets = filter( getActionLog( state ), entry => entry.type === 'ROUTE_SET' );
```
Comment (mcsf) [inline client/state/ui/first-view/selectors.js]: We should be using the proper `ROUTE_SET` constant from `action-types`, rather than the string representation.

While we're here: if lodash's `filter` is only used in its `filter( collection, predicate )` form, then it could be replaced with the native implementation: `getActionLog( state ).filter( entry => … )`. An other usage would be the convenient: `filter( getActionLog( state ), { type: ROUTE_SET } )`.

File: client/state/ui/first-view/test/README.md [hunk 8]
```diff
@@ -0,0 +1,4 @@
+Push Notification Testing
```
Comment (mcsf) [inline client/state/ui/first-view/test/README.md]: Is this title intentional?

**Comment by mcsf** [comment]
> could you have a quick look at the actionLog bits, to make sure we're not going to break anything

Nothing jumps out. This PR merely consumes `getActionLog`, and #6417's intervention to `action-log/reducer` ended up only allowing overrides but not implementing any for `ROUTE_SET` — meaning we keep collecting the same actions that this PR expects.

File: client/state/ui/first-view/test/README.md (hunk 8 above)
Comment (mjuhasz) [inline client/state/ui/first-view/test/README.md]: Good catch, a copy-paste error! Fixed!

File: client/state/ui/first-view/selectors.js [hunk 9]
```diff
@@ -49,3 +58,18 @@ export function shouldViewBeVisible( state ) {
...
 }
+
+function isLikelyInSafariPrivateMode() {
+	// Safari doesn't allow access to Local Storage when in Private Mode;
+	// We will use this check as a proxy to determine whether or not Safari
+	// is in Private Mode
+	let isAbleToAccessLocalStorage = true;
+
+	try {
+		localStorage.checkIfFirstViewPreferencesCanBeCached = null;
```
Comment (ehg) [inline client/state/ui/first-view/selectors.js]: `localStorage` is on the `window` object, so this would break SSR constraints. Do we actually want to test for localStorage here? We use `localForage`, I wonder if there's an established way to check elsewhere in Calypso.

Also, putting this in the selector worries me.

File: client/state/ui/first-view/selectors.js (hunk 7 above)
Comment (mjuhasz) [inline client/state/ui/first-view/selectors.js]: Good point, fixed!

File: client/state/ui/first-view/selectors.js (hunk 9 above)
Comment (mattsherman) [inline client/state/ui/first-view/selectors.js]: @ehg We will pull this back out of the PR and tackle it in a separate PR where we can discuss more and come to the best solution, taking into account things like SSR. Thanks for pointing this out and questioning it!

**Comment by mjuhasz** [comment]
Marking this as Ready to Merge as per feedback (thumbs-up) from reviewers above. Thank you for your contribution!
